config.org
#+TITLE: Doom Emacs Config
#+AUTHOR: Sujal Bajracharya
#+STARTUP: showeverything
#+OPTIONS: toc:2
#+PROPERTY: header-args :tangle config.el

* Table of Contents :toc:
- [[#editor][Editor]]
  - [[#theme][Theme]]
  - [[#font][Font]]
  - [[#cursor][Cursor]]
  - [[#line-numbers][Line Numbers]]
  - [[#tab-size][Tab size]]
  - [[#better-splits][Better Splits]]
- [[#keybindings][Keybindings]]
  - [[#line-start-or-end][Line start or end]]
  - [[#comment-or-uncomment-lines][Comment or Uncomment lines]]
  - [[#word-wrapping-in-org-mode][Word wrapping in org mode]]
  - [[#switch-buffers][Switch buffers]]
- [[#org][Org]]
  - [[#basic-config][Basic Config]]
  - [[#word-wrapping-in-source-blocks][Word wrapping in source blocks]]
  - [[#inline-images][Inline images]]
  - [[#org-agenda][Org Agenda]]
- [[#modeline][Modeline]]
  - [[#doom-modeline-configs][Doom modeline configs]]
  - [[#clock][Clock]]
- [[#dired][Dired]]
  - [[#basic-dired-config][Basic dired config]]
  - [[#evil-dired-keybindings][Evil dired keybindings]]

* Editor
** Theme
#+begin_src emacs-lisp
(setq doom-theme 'doom-one)
#+end_src

** Font
#+begin_src emacs-lisp
(setq doom-font (font-spec :family "JetBrains Mono" :size 20 :weight 'light)
      doom-variable-pitch-font (font-spec :family "Iosevka" :size 15 :weight 'light)
      doom-big-font (font-spec :family "JetBrains Mono" :size 24))

(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))
(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))
#+end_src

#+RESULTS:
| doom--customize-themes-h-28 | doom--customize-themes-h-30 | doom--customize-themes-h-31 |

** Cursor
#+begin_src emacs-lisp
(setq evil-default-cursor       'box
      evil-normal-state-cursor  'box
      evil-emacs-state-cursor   'box
      evil-motion-state-cursor  'box
      evil-insert-state-cursor  'box
      evil-visual-state-cursor  'box
      evil-replace-state-cursor 'box)
#+end_src

** Line Numbers
#+begin_src emacs-lisp
(setq display-line-numbers-type 'relative)
#+end_src

** Tab size
#+begin_src emacs-lisp
(setq-default tab-width 2)
(setq-default evil-shift-width tab-width)
#+end_src

** Better Splits
#+begin_src emacs-lisp
(defun prefer-horizontal-split ()
  (set-variable 'split-height-threshold nil t)
  (set-variable 'split-width-threshold 40 t)) ; make this as low as needed
(add-hook 'markdown-mode-hook 'prefer-horizontal-split)
#+end_src


* Keybindings
** Line start or end
#+begin_src emacs-lisp
(define-key evil-normal-state-map (kbd "H") (kbd "^"))
(define-key evil-normal-state-map (kbd "L") (kbd "$"))
(define-key evil-visual-state-map (kbd "H") (kbd "^"))
(define-key evil-visual-state-map (kbd "L") (kbd "$"))
#+end_src

** Comment or Uncomment lines
#+begin_src emacs-lisp
(map! :leader
      :desc "Comment or uncomment lines" "TAB TAB" #'comment-line)
#+end_src

** Word wrapping in org mode
#+begin_src emacs-lisp
(map! :leader
      :desc "Toggle word wrap in org" "m w" #'toggle-truncate-lines)
#+end_src

** Switch buffers
#+begin_src emacs-lisp
(map! :leader
      :desc "Switch buffers" "b ." #'ivy-switch-buffer)
#+end_src

* Org
** Basic Config
#+begin_src emacs-lisp
(map! :leader
      :desc "Org babel tangle" "m B" #'org-babel-tangle)
(after! org
  (setq inhibit-compacting-font-caches t)
  (custom-set-faces
    '(org-level-1 ((t (:inherit outline-1 :height 1.7))))
    '(org-level-2 ((t (:inherit outline-2 :height 1.6))))
    '(org-level-3 ((t (:inherit outline-3 :height 1.5))))
    '(org-level-4 ((t (:inherit outline-4 :height 1.4))))
    '(org-level-5 ((t (:inherit outline-5 :height 1.3))))
    '(org-level-6 ((t (:inherit outline-6 :height 1.2))))
    '(org-level-7 ((t (:inherit outline-7 :height 1.1))))
    '(org-level-8 ((t (:inherit outline-8 :height 1.0)))))
  (setq org-directory "c:/Users/N I T R O 5/Documents/Org/"
        org-default-notes-file (expand-file-name "notes.org" org-directory)
        org-ellipsis " ▼"
        org-superstar-headline-bullets-list '("◉" "●" "◆" "○" "●" "◆" "○")
        org-superstar-item-bullet-alist '((?+ . ?✦) (?- . ?➤))))
#+end_src

** Word wrapping in source blocks
#+begin_src emacs-lisp
(add-hook 'org-mode-hook
      (lambda ()
        (toggle-truncate-lines nil)))
#+end_src

** Inline images
#+begin_src emacs-lisp
(after! org
  (setq org-startup-with-inline-images t))
#+end_src

** Org Agenda
#+begin_src emacs-lisp
(after! org
  (setq org-agenda-files '("c:/Users/N I T R O 5/Documents/Org/agenda.org")
        org-log-done 'time
        org-log-into-drawer t))
#+end_src

* Modeline
** Doom modeline configs
#+begin_src emacs-lisp
(set-face-attribute 'mode-line nil :font "Cascadia Code-15")
(setq doom-modeline-height 40     ;; sets modeline height
      doom-modeline-bar-width 6   ;; sets right bar width
      doom-modeline-persp-name t  ;; adds perspective name to modeline
      doom-modeline-persp-icon t  ;; adds folder icon next to persp name when set to t
      doom-modeline-buffer-file-name-style 'truncate-upto-project ;; almost full path to file
      doom-modeline-project-detection 'project ;; finds project root folder
      doom-modeline-major-mode-icon nil
      doom-modeline-buffer-modification-icon t ;; removes icon next to file path when file is changed when set to nil
      doom-modeline-highlight-modified-buffer-name nil
      doom-modeline-modal nil
      doom-modeline-modal-icon nil
      doom-modeline-enable-word-count nil)
(after! doom-modeline
  (remove-hook 'doom-modeline-mode-hook #'size-indication-mode) ; filesize in modeline
  (remove-hook 'doom-modeline-mode-hook #'column-number-mode)   ; cursor column in modeline
  (line-number-mode -1)
  (setq doom-modeline-buffer-encoding t))
#+end_src

** Clock
#+begin_src emacs-lisp
(use-package time
  :ensure nil
  :hook (after-init . display-time-mode)
  :custom
  (display-time-default-load-average nil))
#+end_src

* Dired
** Basic dired config
#+begin_src emacs-lisp
(map! :leader
      (:prefix ("d" . "dired")
       :desc "Open dired" "d" #'dired
       :desc "Dired jump to current" "j" #'dired-jump)
      (:after dired
       (:map dired-mode-map
        :desc "Peep-dired image previews" "d p" #'peep-dired
        :desc "Dired view file"           "d v" #'dired-view-file)))
(add-hook 'dired-mode-hook 'all-the-icons-dired-mode)
#+end_src

** Evil dired keybindings
#+begin_src emacs-lisp
(evil-define-key 'normal dired-mode-map
  (kbd "M-RET") 'dired-display-file
  (kbd "h") 'dired-up-directory
  (kbd "l") 'dired-find-file ; use dired-find-file instead of dired-open.
  (kbd "m") 'dired-mark
  (kbd "t") 'dired-toggle-marks
  (kbd "u") 'dired-unmark
  (kbd "C") 'dired-do-copy
  (kbd "D") 'dired-do-delete
  (kbd "J") 'dired-goto-file
  (kbd "M") 'dired-do-chmod
  (kbd "O") 'dired-do-chown
  (kbd "P") 'dired-do-print
  (kbd "R") 'dired-do-rename
  (kbd "T") 'dired-do-touch
  (kbd "Y") 'dired-copy-filenamecopy-filename-as-kill ; copies filename to kill ring.
  (kbd "Z") 'dired-do-compress
  (kbd "+") 'dired-create-directory
  (kbd "-") 'dired-do-kill-lines
  (kbd "% l") 'dired-downcase
  (kbd "% m") 'dired-mark-files-regexp
  (kbd "% u") 'dired-upcase
  (kbd "* %") 'dired-mark-files-regexp
  (kbd "* .") 'dired-mark-extension
  (kbd "* /") 'dired-mark-directories
  (kbd "; d") 'epa-dired-do-decrypt
  (kbd "; e") 'epa-dired-do-encrypt)
#+end_src


config.el
(setq doom-theme 'doom-one)

(setq doom-font (font-spec :family "Iosevka" :size 27 :weight 'light)
      doom-variable-pitch-font (font-spec :family "JetBrains Mono" :size 15 :weight 'light)
      doom-big-font (font-spec :family "Iosevka" :size 30))

(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))
(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))

(setq evil-default-cursor       'box
      evil-normal-state-cursor  'box
      evil-emacs-state-cursor   'box
      evil-motion-state-cursor  'box
      evil-insert-state-cursor  'box
      evil-visual-state-cursor  'box
      evil-replace-state-cursor 'box)

(setq display-line-numbers-type 'relative)

(setq-default tab-width 2)
(setq-default evil-shift-width tab-width)

(defun prefer-horizontal-split ()
  (set-variable 'split-height-threshold nil t)
  (set-variable 'split-width-threshold 40 t)) ; make this as low as needed
(add-hook 'markdown-mode-hook 'prefer-horizontal-split)

(define-key evil-normal-state-map (kbd "H") (kbd "^"))
(define-key evil-normal-state-map (kbd "L") (kbd "$"))
(define-key evil-visual-state-map (kbd "H") (kbd "^"))
(define-key evil-visual-state-map (kbd "L") (kbd "$"))

(map! :leader
      :desc "Comment or uncomment lines" "TAB TAB" #'comment-line)

(map! :leader
      :desc "Toggle word wrap in org" "m w" #'toggle-truncate-lines)

(map! :leader
      :desc "Switch buffers" "b ." #'ivy-switch-buffer)

(map! :leader
      :desc "Toggle tab bar" "t t" #'centaur-tabs-mode)

(map! :leader
      :desc "Org babel tangle" "m C-b" #'org-babel-tangle)
(after! org
  (setq inhibit-compacting-font-caches t)
  (custom-set-faces
    '(org-level-1 ((t (:inherit outline-1 :height 1.6))))
    '(org-level-2 ((t (:inherit outline-2 :height 1.5))))
    '(org-level-3 ((t (:inherit outline-3 :height 1.4))))
    '(org-level-4 ((t (:inherit outline-4 :height 1.3))))
    '(org-level-5 ((t (:inherit outline-5 :height 1.2))))
    '(org-level-6 ((t (:inherit outline-6 :height 1.1))))
    '(org-level-7 ((t (:inherit outline-7 :height 1.0))))
    '(org-level-8 ((t (:inherit outline-8 :height 1.0)))))
  (setq org-directory "c:/Users/N I T R O 5/Documents/Org/"
        org-default-notes-file (expand-file-name "notes.org" org-directory)
        org-ellipsis " ▼"
        org-superstar-headline-bullets-list '("◉" "●" "◆" "○" "●" "◆" "○")
        org-superstar-item-bullet-alist '((?+ . ?✦) (?- . ?➤))))

(add-hook 'org-mode-hook
      (lambda ()
        (toggle-truncate-lines nil)))

(after! org
  (setq org-startup-with-inline-images t))

(after! org
  (setq org-agenda-files '("c:/Users/N I T R O 5/Documents/Org/agenda.org")
        org-log-done 'time
        org-log-into-drawer t))

(setq doom-modeline-height 50
      doom-modeline-bar-width 5   ;; sets right bar width
      doom-modeline-persp-name t  ;; adds perspective name to modeline
      doom-modeline-persp-icon t  ;; adds folder icon next to persp name when set to t
      doom-modeline-buffer-file-name-style 'truncate-upto-project ;; almost full path to file
      doom-modeline-project-detection 'project ;; finds project root folder
      doom-modeline-major-mode-icon nil
      doom-modeline-buffer-modification-icon t ;; removes icon next to file path when file is changed when set to nil
      doom-modeline-highlight-modified-buffer-name nil
      doom-modeline-modal nil
      doom-modeline-modal-icon nil
      doom-modeline-enable-word-count nil)
(after! doom-modeline
  (remove-hook 'doom-modeline-mode-hook #'size-indication-mode) ; filesize in modeline
  (remove-hook 'doom-modeline-mode-hook #'column-number-mode)   ; cursor column in modeline
  (line-number-mode -1)
  (setq doom-modeline-buffer-encoding t))

(use-package time
  :ensure nil
  :hook (after-init . display-time-mode)
  :custom
  (display-time-default-load-average nil))

(use-package battery :ensure nil
  :hook (after-init . display-battery-mode))

(map! :leader
      (:prefix ("d" . "dired")
       :desc "Open dired" "d" #'dired
       :desc "Dired jump to current" "j" #'dired-jump)
      (:after dired
       (:map dired-mode-map
        :desc "Peep-dired image previews" "d p" #'peep-dired
        :desc "Dired view file"           "d v" #'dired-view-file)))
(add-hook 'dired-mode-hook 'nerd-icons-dired-mode)

(evil-define-key 'normal dired-mode-map
  (kbd "M-RET") 'dired-display-file
  (kbd "h") 'dired-up-directory
  (kbd "l") 'dired-find-file ; use dired-find-file instead of dired-open.
  (kbd "m") 'dired-mark
  (kbd "t") 'dired-toggle-marks
  (kbd "u") 'dired-unmark
  (kbd "C") 'dired-do-copy
  (kbd "D") 'dired-do-delete
  (kbd "J") 'dired-goto-file
  (kbd "M") 'dired-do-chmod
  (kbd "O") 'dired-do-chown
  (kbd "P") 'dired-do-print
  (kbd "R") 'dired-do-rename
  (kbd "T") 'dired-do-touch
  (kbd "Y") 'dired-copy-filenamecopy-filename-as-kill ; copies filename to kill ring.
  (kbd "Z") 'dired-do-compress
  (kbd "+") 'dired-create-directory
  (kbd "-") 'dired-do-kill-lines
  (kbd "% l") 'dired-downcase
  (kbd "% m") 'dired-mark-files-regexp
  (kbd "% u") 'dired-upcase
  (kbd "* %") 'dired-mark-files-regexp
  (kbd "* .") 'dired-mark-extension
  (kbd "* /") 'dired-mark-directories
  (kbd "; d") 'epa-dired-do-decrypt
  (kbd "; e") 'epa-dired-do-encrypt)

(map! "C-;" #'smex)
